joomla/joomla-cms
# [5.0][RFC][Events] Make Editor and Buttons plugins to use plugin events

### Summary of Changes

The patch improve Editor and Buttons plugins.
Now they can use a real events.
It similar to #39657 but for Editors and Buttons.

This is fully backward compatible.

**How this work.**
Here I added a new event `onEditorSetup` and an Interface for `EditorProviderInterface` class.
And similar for buttons `onEditorButtonsSetup` and a Button class with Interface `ButtonInterface`.

On `onEditorSetup` an editor plugin should register own `EditorProvider` in `EditorsRegistry`.
On `onEditorButtonsSetup` a button plugin should register own Button(s) in `ButtonsRegistry`.

The system will retrieve a requested Editor from that registry and call required methods, that defined by `EditorProviderInterface`.

Example registering an Editor:
https://github.com/joomla/joomla-cms/blob/059d2b4b96516959dd2cfaafb224c7fe457e7636/plugins/editors/codemirror/src/Extension/Codemirror.php#L50-L55

Example registering an editor Button:
https://github.com/joomla/joomla-cms/blob/059d2b4b96516959dd2cfaafb224c7fe457e7636/plugins/editors-xtd/article/src/Extension/Article.php#L58-L73

### Testing Instructions
Apply patch.
Set default editor to Codemirror and make sure it works.

Testing B/C:
Enable TinyMCE editor and make sure it works.


### Actual result BEFORE applying this Pull Request
Works


### Expected result AFTER applying this Pull Request
Works


### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: https://docs.joomla.org/Plugin/Events
- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/97
